### DISCUSSION TITLE
TorchScript model export used in Andoid app Tensor data float array problem

### DISCUSSION THREAD

**Discussion by atanasko** [ultralytics/ultralytics#737] [Q&A]
I try to use the following code

`
Bitmap resizedBitmap = Bitmap.createScaledBitmap(mBitmap, PrePostProcessor.mInputWidth, PrePostProcessor.mInputHeight, true);
`

`
final Tensor inputTensor = TensorImageUtils.bitmapToFloat32Tensor(resizedBitmap, PrePostProcessor.NO_MEAN_RGB, PrePostProcessor.NO_STD_RGB);
`

`
 IValue inputs = IValue.from(inputTensor);
`

`
Tensor outputTensor = mModule.forward(inputs).toTensor();
`

`
 final float[] outputs = outputTensor.getDataAsFloatArray();
`

to produce the output array to be able to draw the bboxes, but my result array does not look that contains the correct data. From the doc 

> The output of an object detector is a set of bounding boxes that enclose the objects in the image, along with class labels and confidence scores for each box.

but the output array look like it contains only bboxes (not sure that they are correct).

Model was custom trained model and exported with 

`
yolo export model="./runs/detect/train16/weights/best.pt" format=torchscript
`

When I test the model with 

`
yolo detect predict model="./runs/detect/train16/weights/best.pt" source="rocket_launch.mp4" show=True
`

all works fine.

Android Pytorch dependencies

`
implementation 'org.pytorch:pytorch_android:1.12'
`
`
implementation 'org.pytorch:pytorch_android_torchvision:1.12'
`

Any advice how to check is export script is valid?

**Comment by atanasko** [comment]
I test the model using

`
yolo detect predict model="./runs/detect/train16/weights/best.torchscript" source="rocket_launch.mp4" show=True
`

but still can not retrieve the correct data in the Android app :-(

Any advice is appreciated!

**Reply by glenn-jocher** [reply]
@atanasko it sounds like you're experiencing issues with the output format of your TorchScript model when running it on an Android app. The output tensor from a YOLO model typically includes both the bounding box coordinates and the class confidences, but it seems you're only seeing part of the expected output.

First, ensure that your model export to TorchScript was successful and that the model is functioning correctly by using the `predict` mode with the TorchScript model on your desktop environment. If the predictions are correct there, the export is likely fine.

Next, on the Android side, verify that you're interpreting the output tensor correctly. The output tensor's shape and contents can vary depending on the model, so you'll need to understand the specific output format for YOLOv8. This format typically includes bounding box coordinates, confidence scores, and class predictions.

If the desktop test is successful but you're still facing issues on Android, it could be related to how you're processing the output tensor in the app. Make sure you're reshaping or slicing the tensor as needed to extract all the relevant information.

For further guidance on the expected output format and how to handle it, please refer to the documentation on the `predict` mode and the output tensor structure. If you continue to have issues, you might want to check for any version mismatches between the PyTorch versions used for exporting the model and the PyTorch Android dependencies, as these can sometimes cause discrepancies in behavior.

Remember, the community and the Ultralytics team are here to help, so don't hesitate to reach out if you need more assistance. Good luck with your application development! 🚀

**Comment by atanasko** [comment]
When I inspect the result Tensor and the float array from the Tensor object I have a Tensor [1, 8, 8400] and to me look like expected because I have 4 classes and I expect the other values to be score for every particular class. But in the result float array to me it look like first bboxes (x, y, w, h) in the first half of the array and in the second half of the array scores are added, as in the case below (index in brackets)

`
[33348]  bb : 503.1715, 507.16373, 505.32724, 505.98315 [66948]  score: 1.365644E-4, 0.0014117783, 0.45453435, 0.13102075
`
`
[33368]  bb : 498.7717, 505.04846, 503.06573, 499.7424 [66968]  score: 3.189207E-4, 0.008651144, 0.64653474, 0.36695477
`
`
[33388]  bb : 496.12158, 501.20477, 499.07532, 500.95465 [66988]  score: 4.076895E-4, 0.0055294246, 0.4945656, 0.24356054
`
`
[33408]  bb : 504.01505, 513.1348, 512.05286, 516.615 [67008]  score: 3.494384E-4, 0.0032577545, 0.33719948, 0.14475583
`
`
[33428]  bb : 392.64484, 466.43222, 482.61945, 487.92276 [67028]  score: 5.5740704E-4, 0.013573675, 0.5446357, 0.36364305
`
`
[33448]  bb : 379.116, 388.1064, 403.084, 421.47614 [67048]  score: 3.7921825E-4, 0.013157838, 0.57693213, 0.2928344
`
`
[33468]  bb : 371.716, 382.32294, 369.3178, 368.37723 [67068]  score: 8.664998E-5, 0.0011737087, 0.3274485, 0.037356436
`

Do I miss something here?

**Reply by glenn-jocher** [reply]
@atanasko it seems like you're encountering an issue with interpreting the output tensor from your TorchScript model on Android. The output tensor shape `[1, 8, 8400]` suggests that for each of the 8400 grid cells, the model outputs 8 values. These typically include 4 for the bounding box (x, y, width, height), 1 for the objectness score, and the remaining for class confidence scores.

The output tensor is usually structured in a way that the bounding box coordinates and class scores are interleaved or concatenated in a specific format. It's important to understand the exact structure of the output tensor to correctly parse and use this data.

To verify if the export script worked correctly and the model is exporting the expected outputs, you can:

1. Use the `yolo export` command with the `--simplify` flag to ensure the model is exported in a simplified format that might be more compatible with the Android environment.
2. Compare the output of the TorchScript model with the output of the `.pt` model on the same input data using a Python environment to ensure consistency.

If the outputs match, then the issue might be with how you're interpreting the output tensor in your Android application. You may need to adjust your parsing logic to correctly extract the bounding boxes and class scores from the output tensor based on its structure.

Remember that the output tensor's structure is determined by the model's architecture and the export process, so it's crucial to have a clear understanding of that structure when developing your parsing logic. If necessary, consult the documentation at https://docs.ultralytics.com for more details on the output format and how to handle it post-inference.

**Reply by atanasko** [reply]
@glenn-jocher Thank you!

**Reply by glenn-jocher** [reply]
@atanasko you're welcome!
